README.md
# Tracing - Datadog Application Performance Monitoring (APM)
`tracing-datadog-apm` uses the [tracing](https://github.com/tokio-rs/tracing) crate
to produce application traces in a format usable by Datadog's APM service.

From the [tracing](https://github.com/tokio-rs/tracing) README:

> `tracing` is a framework for instrumenting Rust programs to collect
> structured, event-based diagnostic information. `tracing` is maintained by the
> Tokio project, but does _not_ require the `tokio` runtime to be used.

By using the [tracing](https://github.com/tokio-rs/tracing) crate
we are able to produce structured trace information in asynchronous and in
multi-threaded environments, which are exactly the kinds of environments
typically being monitored with Datadog APM.

## Installation
Add this line to your `Cargo.toml`:
```toml
tracing-datadog-apm = "0.0.1"
```

That will provide an implementation of `tracing::Subscriber` that will pick up 
trace data, as well as a Datadog client equipped to pass trace data to the 
Datadog agent running on the machine.

If you want APM instrumentation out of the box, this crate aims to provide a few
common ones.  Each instrumentation can be enabled as a feature.  Right now the crate
provides:

`actix-web` middleware
```toml
tracing-datadog-apm = { version = "0.0.1", features = ["actix-web"] }
```

## Usage
### 1) Setup Datadog Agent
First, make sure your application also has a Datadog agent running in the background.
See [more info here](https://docs.datadoghq.com/agent/).

### 2) Create Datadog client
Then, in your code, create a Datadog client. If you're using the latest version of the
Datadog agent, and the agent is running on the same machine as your application,
you can use `.create_default()` to make a client with the correct values.
```rust
let datadog_client = tracing_datadog_apm::datadog_client::Client::create_default();
```

If your Datadog agent is running on another machine or on a different port than the default
8126, you can provide your own config by using `.create_with_config()` and passing in an instance
of `ClientConfig`. `ClientConfig` lets you chain method calls to set the fields you want.
```rust
let datadog_client = tracing_datadog_apm::datadog_client::Client::create_with_config(
    tracing_datadog_apm::datadog_client::ClientConfig::new()
        .datadog_agent_host("foo")
        .datadog_agent_port(1234),
);
```

### 3) Create Datadog tracing `Subscriber`
Next, create a Datadog tracing `Subscriber`. This will take ownership of the Datadog
client created in the previous step. It will also take a `TracingSubscriberDatadogConfig`,
which maps span names to service names and span type. 

The span name is the name of the span according to the `tracing` library. 
The service name and span type are properties of the span when it gets into Datadog 
(notice that both `tracing` and Datadog have a notion of a span).
```rust
use tracing_datadog_apm::subscriber::{TracingSubscriberDatadog, TracingSubscriberDatadogConfig};

let datadog_tracing_subscriber = TracingSubscriberDatadog::new(
    datadog_client,
    TracingSubscriberDatadogConfig::new()
        .add_mapping(
            SpanName("http.request"),
            (ServiceName("my-service-rest"), SpanType::Web),
        )
        .add_mapping(
            SpanName("database.query"),
            (ServiceName("my-service-database"), SpanType::Db),
        )
        .add_mapping(
            SpanName("redis.request"),
            (ServiceName("my-service-cache"), SpanType::Cache),
        )
        .add_mapping(
            SpanName("custom.function"),
            (ServiceName("my-service-function"), SpanType::Custom),
        ),
);
```

There are 4 available `SpanType`s defined by Datadog: `Web`, `Db`, `Cache`, and
`Custom`.

Note that each service name provided will show up as a separate service in the
[APM dashboard](https://app.datadoghq.com/apm), and any time a trace from one
service touches multiple services, each service will show up in the latency breakdown
graph with a percentage of time spent in that service.

Example: See the `% of Time Spent by Service` graph in the lower right.
![Datadog example graphs](https://datadog-docs.imgix.net/images/tracing/visualization/service/out_of_the_box_service_graph.b7b972a596bbec97c174fcd132cb2539.png?fit=max&auto=format&w=1792&h=1009&dpr=2)

You can also map multiple span names and span types to the same service name. In that case that service will show
up in APM, but it will have a dropdown that allows you to select the spans you want to see displayed.

### 4) Set the Datadog Subscriber as the global subscriber
In your application, you will need to be using the [tracing](https://github.com/tokio-rs/tracing)
crate. You should set the Datadog `Subscriber` as the global subscriber (in future
iterations it will also be available as a `Layer` so it can be combined with other
`Subscriber`s).

```rust
tracing::subscriber::set_global_default(datadog_tracing_subscriber)
        .expect("Setting tracing default failed");
```

### 5) Instrument everything
The easiest way to instrument things in an async, multi-threaded environment is 
to use the `tracing` `instrument` attribute macro.

```rust
#[instrument(name = "call.other_service", skip_all, fields(resource, http_method, http_url, http_status_code, error_msg))]
async fn call_other_service(user_uuid: uuid::Uuid) -> bool {
    if let Ok(host) = env::fetch("OTHER_SERVICE_HOST") {
        // define the URL and method for our REST request
        let url = format!("{}/users/{}", host, user_uuid.to_hyphenated());
        let method = "GET";

        // Record some data in the current span about this request:
        // (the current span being a `call.other_service` span created 
        // when we enter this function, thanks to that `instrument` tag)
        //
        // - `resource` is telling us what specifically is being hit in this span.
        // It will show up (along with the other `resource` types) on the APM 
        // dashboard for spans of the type `call.other_service`
        //
        // - `http.method` and `http.url` are just some metadata we're collecting
        //
        let current_span = tracing::Span::current();
        current_span.record("resource", &*format!("{} {}", method, "Other Service"));
        current_span.record("http_method", &*method);
        current_span.record("http_url", &*url);

        // Make the REST request (this stuff is arbitrary, I'm just using
        // reqwest as an example)
        let client = reqwest::Client::new();
        match client.get(&url).send().await {
            Ok(resp) => {
                // Record the response status code as some more metadata
                let current_span = tracing::Span::current();
                current_span.record("http.status_code", resp.status().as_str());
                resp.status().is_success()
            }
            Err(e) => {
                // Alternately, if the request fails, record the error message
                let err_msg = format!("Could not call other service with error: {}", e);
                let current_span = tracing::Span::current();
                current_span.record("error.msg", &*err_msg);
                log::error!("{}", err_msg);
                false
            }
        }
    } else {
        // Record an error message if we failed getting the env var
        let err_msg = "Failed to fetch OTHER_SERVICE_HOST env var";
        let current_span = tracing::Span::current();
        current_span.record("error.msg", err_msg);
        log::error!("{}", err_msg);
        false
    }
}
```

To `instrument` properly, pick a name for the span (this will map to a service and
span type, as defined in the `TracingDatadogSubscriberConfig` - if it's not set in the
config, it will be ignored).  Then `skip` the function args (they will be passed
along by default, unless you skip them). Then define what fields you would actually like to pass along.

The acceptable field names are:
* `trace_id` - the id of the current trace - normally does not need to be passed explicitly
* `parent_id` - the id of the parent span - normally does not need to be passed explicitly
* `resource` - resource name within the given span
* `start` - the start time in nanos from the Unix epoch - normally doesn't need to be passed explicitly
* `http_method` - metadata for http requests (in or out)
* `http_url` - metadata for http requests (in or out)
* `http_status_code` - metadata for http requests (in or out)
* `error_type` - type of the error that occurred (a string)
* `error_msg` - accompanying error message
* `error_stack` - the whole error stack if you have it as a string

At bare minimum, all spans should have a `resource`. For `Web` spans this is easy:
what's the resource for the REST request?  For a `Db` span it is usually the
SQL query but with placeholder values, i.e. `SELECT $1 FROM table WHERE id = $2;`.
`Cache` spans will also often have a query that can serve as the `resource`.
`Custom` spans can do whatever they'd like.

`Web` spans will also typically include the multiple `http` and `error` parameters.

The other span types can make use of the `error` parameters if they need.

For more information on spans, check out 
[these docs](https://tracing-rs.netlify.app/tracing/index.html#spans) 
and for the `instrument` attribute macro, 
[look here](https://tracing-rs.netlify.app/tracing/attr.instrument.html).

For more information on Datadog's treatment of traces and spans,
refer to their [documentation on tracing](https://docs.datadoghq.com/tracing/guide).


CHANGELOG.md
# Changelog
This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.1] - 2023-08-10
### Added
- Configurable Datadog agent client with MPSC channel for sending traces
- A `tokio-tracing` `Subscriber` to record trace data in Datadog's APM format
- Optional `actix-web` middleware to trace all requests if you're using the `actix-web` server framework

[0.0.1]: https://github.com/kraig-mcfadden/tracing-datadog-apm-rust/releases/tag/v0.0.1


tests/datadog_client_integration_tests.rs
use tracing_datadog_apm::datadog_client::{Client, ClientConfig, Traces};

#[test]
fn test_create_datadog_client_default() {
    // given
    let client = Client::create_default();

    // when
    client.send_traces(Traces::new()); // no output here, just checking it didn't panic
}

#[test]
fn test_create_datadog_client_with_config() {
    // given
    let config = ClientConfig::new().datadog_agent_host("foo");
    let client = Client::create_with_config(config);

    // when
    client.send_traces(Traces::new()); // no output here, just checking it didn't panic
}

#[test]
fn test_create_datadog_client_with_config_chained_setters() {
    // given
    let client = Client::create_with_config(
        ClientConfig::new()
            .datadog_agent_host("foo")
            .datadog_agent_port(1234),
    );

    // when
    client.send_traces(Traces::new()); // no output here, just checking it didn't panic
}


src/instrumentation_actix_web.rs
use actix_web::body::MessageBody;
use actix_web::dev::{Service, ServiceRequest, ServiceResponse, Transform};
use actix_web::Error;
use futures::future::{ok, Ready};
use std::fmt::{Debug, Formatter};
use std::future::Future;
use std::pin::Pin;
use std::str::FromStr;
use std::time::{SystemTime, UNIX_EPOCH};
use tracing::instrument;

type PinnedBoxedFuture<T> = Pin<Box<dyn Future<Output = T>>>;

#[derive(Debug)]
pub struct ActixDatadogTracer;

impl<S, B> Transform<S, ServiceRequest> for ActixDatadogTracer
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
    S::Future: 'static,
    B: 'static + MessageBody,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type Transform = ActixDatadogTracerMiddleware<S>;
    type InitError = ();
    type Future = Ready<Result<Self::Transform, Self::InitError>>;

    fn new_transform(&self, service: S) -> Self::Future {
        ok(ActixDatadogTracerMiddleware { service })
    }
}

pub struct ActixDatadogTracerMiddleware<S> {
    service: S,
}

impl<S> Debug for ActixDatadogTracerMiddleware<S> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str("ActixDatadogTracerMiddleware")
    }
}

impl<S, B> Service<ServiceRequest> for ActixDatadogTracerMiddleware<S>
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
    S::Future: 'static,
    B: 'static + MessageBody,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type Future = PinnedBoxedFuture<Result<Self::Response, Self::Error>>;

    actix_service::forward_ready!(service);

    #[instrument(
        name = "request",
        skip_all,
        fields(
            trace_id,
            parent_id,
            resource,
            start,
            http_method,
            http_url,
            http_status_code,
            error_type,
            error_msg,
            error_stack
        )
    )]
    fn call(&self, req: ServiceRequest) -> Self::Future {
        let recordable_data = extract_recordable_data(&req);

        let fut = self.service.call(req);

        Box::pin(async move {
            let current_span = tracing::Span::current();
            current_span.record(
                "resource",
                &*format!("{} {}", recordable_data.method, recordable_data.resource),
            );
            current_span.record("http.method", &*recordable_data.method);
            current_span.record("http.url", &*recordable_data.url);

            if let Some(start) = recordable_data.maybe_start {
                current_span.record("start", start);
            }
            if let Some(trace_id) = recordable_data.maybe_trace_id {
                current_span.record("trace_id", trace_id);
            }
            if let Some(parent_id) = recordable_data.maybe_parent_id {
                current_span.record("parent_id", parent_id);
            }

            let res = fut.await?;

            let current_span = tracing::Span::current();
            current_span.record("http.status_code", res.status().as_str());

            if res.status().is_server_error() {
                current_span.record(
                    "error.msg",
                    &*format!(
                        "Request has failed with HTTP error: {}",
                        res.status().as_str()
                    ),
                );
                if let Some(err) = res.response().error() {
                    current_span.record("error.type", &*format!("{:?}", err));
                    current_span.record("error.stack", &*format!("{:?}", err));
                } else {
                    current_span.record("error.type", "Server side error");
                }
            }

            Ok(res)
        })
    }
}

struct RecordableData {
    maybe_start: Option<u64>,
    resource: String,
    method: String,
    url: String,
    maybe_trace_id: Option<u64>,
    maybe_parent_id: Option<u64>,
}

#[inline]
fn extract_recordable_data(req: &ServiceRequest) -> RecordableData {
    let (maybe_trace_id, maybe_parent_id) = extract_trace_and_parent(req);
    RecordableData {
        maybe_start: SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .ok()
            .map(|duration| duration.as_nanos() as u64),
        resource: req.match_pattern().unwrap_or_else(|| String::from("404")),
        method: req.method().to_string(),
        url: req.uri().to_string(),
        maybe_trace_id,
        maybe_parent_id,
    }
}

#[inline]
fn extract_trace_and_parent(req: &ServiceRequest) -> (Option<u64>, Option<u64>) {
    let mut maybe_trace_id = None;
    let mut maybe_parent_id = None;
    req.headers().iter().for_each(|(key, value)| {
        match &*key.as_str().trim().to_lowercase() {
            // Datadog headers
            "x-datadog-trace-id" => value.to_str().map(|s| maybe_trace_id = Some(s)),
            "x-datadog-parent-id" => value.to_str().map(|s| maybe_parent_id = Some(s)),

            // Zipkin B3 headers
            "x-b3-traceid" => value.to_str().map(|s| maybe_trace_id = Some(s)),
            "x-b3-spanid" => value.to_str().map(|s| maybe_parent_id = Some(s)),

            // B3 single header
            "b3" => {
                // b3: {TraceId}-{SpanId}-{SamplingState}-{ParentSpanId}
                value.to_str().map(|s| {
                    let parts: Vec<&str> = s.split('-').collect();
                    if parts.len() >= 2 {
                        maybe_trace_id = Some(parts[0]);
                        maybe_parent_id = Some(parts[1]);
                    }
                })
            }

            _ => Ok(()),
        }
        .ok();
    });
    (
        maybe_trace_id.and_then(|s| u64::from_str(s).ok()),
        maybe_parent_id.and_then(|s| u64::from_str(s).ok()),
    )
}


src/datadog_client.rs
use rand::Rng;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fmt::Debug;
use std::num::NonZeroU64;
use std::str::FromStr;
use std::sync::mpsc::Sender;
use std::sync::Mutex;
use std::thread::JoinHandle;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

/// ClientConfig comes with sensible defaults. Calling either ClientConfig::default() or
/// ClientConfig::new() will create a ClientConfig instance with these defaults. If any
/// fields need to be changed, you only need to call the setter for that field - the others
/// will retain their default value.
///
/// If you intend to use entirely default values, you don't need to instantiate a ClientConfig.
/// Instead, you can just call Client::create_default().
pub struct ClientConfig {
    datadog_agent_host: String,
    datadog_agent_port: u32,
    connect_timeout_ms: u64,
    request_timeout_ms: u64,
}

impl ClientConfig {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn datadog_agent_host(mut self, host: impl Into<String>) -> Self {
        self.datadog_agent_host = host.into();
        self
    }

    pub fn datadog_agent_port(mut self, port: u32) -> Self {
        self.datadog_agent_port = port;
        self
    }

    pub fn connect_timeout_ms(mut self, ms: u64) -> Self {
        self.connect_timeout_ms = ms;
        self
    }

    pub fn request_timeout_ms(mut self, ms: u64) -> Self {
        self.request_timeout_ms = ms;
        self
    }
}

impl Default for ClientConfig {
    fn default() -> Self {
        Self {
            datadog_agent_host: String::from("localhost"),
            datadog_agent_port: 8126,
            connect_timeout_ms: 100,
            request_timeout_ms: 100,
        }
    }
}

pub struct Client {
    sender_mutex: Mutex<Sender<serde_json::Value>>,
    _daemon: JoinHandle<()>,
}

impl Client {
    pub fn create_default() -> Self {
        Self::create_with_config(ClientConfig::default())
    }

    pub fn create_with_config(config: ClientConfig) -> Self {
        let (sender, receiver) = std::sync::mpsc::channel::<serde_json::Value>();

        let daemon: JoinHandle<()> = std::thread::spawn(move || {
            log::info!("Starting daemon thread to pass traces to Datadog agent");
            let client = reqwest::blocking::ClientBuilder::new()
                .connect_timeout(Duration::from_millis(config.connect_timeout_ms))
                .timeout(Duration::from_millis(config.request_timeout_ms))
                .build()
                .map_err(|e| log::error!("Failed to construct client, killing daemon; err {:?}", e))
                .unwrap();
            let dd_agent_url = format!(
                "http://{}:{}/v0.3/traces",
                config.datadog_agent_host, config.datadog_agent_port
            );
            loop {
                match receiver.recv() {
                    Ok(trace_json) => {
                        send_traces_to_datadog_agent(&client, &dd_agent_url, trace_json);
                    }
                    Err(e) => log::error!("Failed to receive traces on mpsc channel; err {:?}", e),
                }
            }
        });

        Self {
            sender_mutex: Mutex::new(sender),
            _daemon: daemon,
        }
    }

    #[inline]
    pub fn send_traces(&self, traces: Traces) {
        let trace_json = serde_json::to_value(traces).unwrap_or_else(|e| {
            log::error!("Failed to serialize traces into JSON value. Err: {}", e);
            serde_json::Value::default()
        });
        match self.sender_mutex.lock() {
            Ok(sender) => match sender.send(trace_json) {
                Ok(_) => {}
                Err(e) => log::error!("Failed to send traces on mpsc channel; err {:?}", e),
            },
            Err(e) => log::error!("Failed to get lock on sender; err {:?}", e),
        }
    }
}

#[inline]
fn send_traces_to_datadog_agent(
    client: &reqwest::blocking::Client,
    dd_agent_url: &str,
    trace_json: serde_json::Value,
) {
    match client.put(dd_agent_url).body(trace_json.to_string()).send() {
        Ok(resp) => log::debug!(
            "Successfully sent trace to Datadog agent; response: {:?}",
            resp
        ),
        Err(e) => log::error!("Failed to send trace to Datadog agent; error: {}", e),
    };
}

pub type Traces = Vec<Trace>;

pub type Trace = Vec<Span>;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Span {
    duration: u64,
    error: u32,
    meta: HashMap<String, String>,
    metrics: HashMap<String, u64>,
    name: &'static str,
    parent_id: Option<u64>,
    resource: String,
    service: &'static str,
    span_id: u64,
    start: u64,
    trace_id: u64,
    r#type: &'static str,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum SpanType {
    Web,
    Db,
    Cache,
    Custom,
}

impl SpanType {
    #[inline]
    pub fn as_str(&self) -> &'static str {
        match *self {
            SpanType::Web => "web",
            SpanType::Db => "db",
            SpanType::Cache => "cache",
            SpanType::Custom => "custom",
        }
    }
}

impl FromStr for SpanType {
    type Err = ();

    #[inline]
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match &*s.trim().to_lowercase() {
            "web" => SpanType::Web,
            "db" => SpanType::Db,
            "cache" => SpanType::Cache,
            "custom" => SpanType::Custom,
            _ => SpanType::Custom,
        })
    }
}

#[derive(Copy, Clone, Debug)]
pub struct ServiceName(pub &'static str);

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
pub struct SpanName(pub &'static str);

#[derive(Copy, Clone, Debug)]
pub enum SpanMetaKey {
    Service,
    Env,
    Version,
    HttpMethod,
    HttpUrl,
    HttpStatusCode,
    ErrorType,
    ErrorMsg,
    ErrorStack,
}

impl std::fmt::Display for SpanMetaKey {
    #[inline]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Service => f.write_str("service"),
            Self::Env => f.write_str("env"),
            Self::Version => f.write_str("version"),
            Self::HttpMethod => f.write_str("http.method"),
            Self::HttpUrl => f.write_str("http.url"),
            Self::HttpStatusCode => f.write_str("http.status_code"),
            Self::ErrorType => f.write_str("error.type"),
            Self::ErrorMsg => f.write_str("error.msg"),
            Self::ErrorStack => f.write_str("error.stack"),
        }
    }
}

#[derive(Debug, Clone)]
pub struct SpanBuilder {
    error: bool,
    meta: HashMap<String, String>,
    metrics: HashMap<String, u64>,
    name: SpanName,
    pub parent_id: Option<NonZeroU64>,
    resource: String,
    service: ServiceName,
    pub span_id: NonZeroU64,
    start: SystemTime,
    pub trace_id: NonZeroU64,
    r#type: SpanType,
}

impl Default for SpanBuilder {
    #[inline]
    fn default() -> Self {
        Self {
            error: false,
            meta: HashMap::new(),
            metrics: HashMap::new(),
            name: SpanName(""),
            parent_id: None,
            resource: String::new(),
            service: ServiceName(""),
            span_id: generate_id(),
            start: SystemTime::now(),
            trace_id: generate_id(),
            r#type: SpanType::Custom,
        }
    }
}

impl SpanBuilder {
    #[inline]
    pub fn name(&mut self, name: SpanName) -> &mut Self {
        self.name = name;
        self
    }

    #[inline]
    pub fn service(&mut self, service: ServiceName) -> &mut Self {
        self.service = service;
        self
    }

    #[inline]
    pub fn resource(&mut self, resource: String) -> &mut Self {
        self.resource = resource;
        self
    }

    #[inline]
    pub fn span_type(&mut self, span_type: SpanType) -> &mut Self {
        self.r#type = span_type;
        self
    }

    #[inline]
    pub fn trace_id(&mut self, trace_id: NonZeroU64) -> &mut Self {
        self.trace_id = trace_id;
        self
    }

    #[inline]
    pub fn start(&mut self, start: SystemTime) -> &mut Self {
        self.start = start;
        self
    }

    #[inline]
    pub fn error(&mut self, error: bool) -> &mut Self {
        self.error = error;
        self
    }

    #[inline]
    pub fn add_meta(&mut self, key: SpanMetaKey, value: impl Into<String>) -> &mut Self {
        self.meta.insert(key.to_string(), value.into());
        self
    }

    #[inline]
    pub fn metrics(&mut self, metrics: HashMap<String, u64>) -> &mut Self {
        self.metrics = metrics;
        self
    }

    #[inline]
    pub fn parent_id(&mut self, parent_id: NonZeroU64) -> &mut Self {
        self.parent_id = Some(parent_id);
        self
    }

    #[inline]
    pub fn build(&self) -> Span {
        let duration = SystemTime::now()
            .duration_since(self.start)
            .unwrap_or_else(|_| Duration::from_nanos(0))
            .as_nanos() as u64;
        Span {
            duration,
            error: if self.error { 1 } else { 0 },
            meta: self.meta.clone(),
            metrics: self.metrics.clone(),
            name: self.name.0,
            parent_id: self.parent_id.map(NonZeroU64::get),
            resource: self.resource.clone(),
            service: self.service.0,
            span_id: self.span_id.get(),
            start: self
                .start
                .duration_since(UNIX_EPOCH)
                .expect("Time went backwards")
                .as_nanos() as u64,
            trace_id: self.trace_id.get(),
            r#type: self.r#type.as_str(),
        }
    }
}

#[inline]
pub fn generate_id() -> NonZeroU64 {
    rand::thread_rng().gen()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_config() {
        let config = ClientConfig::default();
        assert_eq!(config.datadog_agent_host, "localhost");
        assert_eq!(config.datadog_agent_port, 8126);
        assert_eq!(config.connect_timeout_ms, 100);
        assert_eq!(config.request_timeout_ms, 100);
    }

    #[test]
    fn test_new_config() {
        let config = ClientConfig::new();
        assert_eq!(config.datadog_agent_host, "localhost");
        assert_eq!(config.datadog_agent_port, 8126);
        assert_eq!(config.connect_timeout_ms, 100);
        assert_eq!(config.request_timeout_ms, 100);
    }

    #[test]
    fn test_config_host() {
        let config = ClientConfig::new().datadog_agent_host("foo");
        assert_eq!(config.datadog_agent_host, "foo");
        assert_eq!(config.datadog_agent_port, 8126);
        assert_eq!(config.connect_timeout_ms, 100);
        assert_eq!(config.request_timeout_ms, 100);
    }

    #[test]
    fn test_config_port() {
        let config = ClientConfig::new().datadog_agent_port(1234);
        assert_eq!(config.datadog_agent_host, "localhost");
        assert_eq!(config.datadog_agent_port, 1234);
        assert_eq!(config.connect_timeout_ms, 100);
        assert_eq!(config.request_timeout_ms, 100);
    }

    #[test]
    fn test_config_connect_timeout() {
        let config = ClientConfig::new().connect_timeout_ms(500);
        assert_eq!(config.datadog_agent_host, "localhost");
        assert_eq!(config.datadog_agent_port, 8126);
        assert_eq!(config.connect_timeout_ms, 500);
        assert_eq!(config.request_timeout_ms, 100);
    }

    #[test]
    fn test_config_request_timeout() {
        let config = ClientConfig::new().request_timeout_ms(750);
        assert_eq!(config.datadog_agent_host, "localhost");
        assert_eq!(config.datadog_agent_port, 8126);
        assert_eq!(config.connect_timeout_ms, 100);
        assert_eq!(config.request_timeout_ms, 750);
    }

    #[test]
    fn test_config_chain() {
        let config = ClientConfig::new()
            .datadog_agent_host("foo")
            .datadog_agent_port(1234)
            .connect_timeout_ms(500)
            .request_timeout_ms(750);
        assert_eq!(config.datadog_agent_host, "foo");
        assert_eq!(config.datadog_agent_port, 1234);
        assert_eq!(config.connect_timeout_ms, 500);
        assert_eq!(config.request_timeout_ms, 750);
    }

    #[test]
    fn test_span_type_web() {
        let span_type = SpanType::from_str("web").unwrap();
        assert_eq!(span_type, SpanType::Web);
        assert_eq!(span_type.as_str(), "web");
    }

    #[test]
    fn test_span_type_db() {
        let span_type = SpanType::from_str("db").unwrap();
        assert_eq!(span_type, SpanType::Db);
        assert_eq!(span_type.as_str(), "db");
    }

    #[test]
    fn test_span_type_cache() {
        let span_type = SpanType::from_str("cache").unwrap();
        assert_eq!(span_type, SpanType::Cache);
        assert_eq!(span_type.as_str(), "cache");
    }

    #[test]
    fn test_span_type_custom() {
        let span_type = SpanType::from_str("custom").unwrap();
        assert_eq!(span_type, SpanType::Custom);
        assert_eq!(span_type.as_str(), "custom");
    }

    #[test]
    fn test_span_type_from_str_capitalized() {
        let span_type = SpanType::from_str("Web").unwrap();
        assert_eq!(span_type, SpanType::Web);
        assert_eq!(span_type.as_str(), "web");
    }

    #[test]
    fn test_span_type_other_string_defaults_to_custom() {
        let span_type = SpanType::from_str("fake type").unwrap();
        assert_eq!(span_type, SpanType::Custom);
        assert_eq!(span_type.as_str(), "custom");
    }

    #[test]
    fn test_default_span_builder() {
        let span = SpanBuilder::default().build();
        assert_eq!(span.error, 0);
        assert_eq!(span.meta, HashMap::new());
        assert_eq!(span.metrics, HashMap::new());
        assert_eq!(span.name, "");
        assert_eq!(span.parent_id, None);
        assert_eq!(span.resource, "");
        assert_eq!(span.service, "");
        assert!(span.span_id > 0);
        assert!(span.start > 0);
        assert!(span.trace_id > 0);
        assert_eq!(span.r#type, "custom");
    }

    #[test]
    fn test_span_builder() {
        let start = SystemTime::now()
            .checked_sub(Duration::from_nanos(100))
            .unwrap();
        let parent_id = NonZeroU64::new(5).unwrap();
        let trace_id = NonZeroU64::new(100).unwrap();
        let name = "foo";
        let resource = "bar";
        let service = "aliceandbob";
        let r#type = SpanType::Db;
        let span = SpanBuilder::default()
            .start(start)
            .parent_id(parent_id)
            .trace_id(trace_id)
            .error(true)
            .name(SpanName(name))
            .resource(String::from(resource))
            .service(ServiceName(service))
            .span_type(r#type)
            .build();
        assert!(span.duration > 100);
        assert_eq!(span.error, 1);
        assert_eq!(span.meta, HashMap::new());
        assert_eq!(span.metrics, HashMap::new());
        assert_eq!(span.name, name);
        assert_eq!(span.parent_id, Some(parent_id.get()));
        assert_eq!(span.resource, resource);
        assert_eq!(span.service, service);
        assert!(span.span_id > 0);
        assert_eq!(
            span.start,
            start.duration_since(UNIX_EPOCH).unwrap().as_nanos() as u64
        );
        assert_eq!(span.trace_id, trace_id.get());
        assert_eq!(span.r#type, "db");
    }

    #[test]
    fn test_span_meta_key_service() {
        assert_eq!(&*SpanMetaKey::Service.to_string(), "service");
    }

    #[test]
    fn test_span_meta_key_env() {
        assert_eq!(&*SpanMetaKey::Env.to_string(), "env");
    }

    #[test]
    fn test_span_meta_key_version() {
        assert_eq!(&*SpanMetaKey::Version.to_string(), "version");
    }

    #[test]
    fn test_span_meta_key_http_method() {
        assert_eq!(&*SpanMetaKey::HttpMethod.to_string(), "http.method");
    }

    #[test]
    fn test_span_meta_key_http_url() {
        assert_eq!(&*SpanMetaKey::HttpUrl.to_string(), "http.url");
    }

    #[test]
    fn test_span_meta_key_http_status_code() {
        assert_eq!(
            &*SpanMetaKey::HttpStatusCode.to_string(),
            "http.status_code"
        );
    }

    #[test]
    fn test_span_meta_key_error_msg() {
        assert_eq!(&*SpanMetaKey::ErrorMsg.to_string(), "error.msg");
    }

    #[test]
    fn test_span_meta_key_error_stack() {
        assert_eq!(&*SpanMetaKey::ErrorStack.to_string(), "error.stack");
    }

    #[test]
    fn test_span_meta_key_error_type() {
        assert_eq!(&*SpanMetaKey::ErrorType.to_string(), "error.type");
    }
}
